### PR TITLE
Jcurtisk/exclude tests c++runtime

### PR DESCRIPTION
**Pull request by jcurtisk**
## What does this PR do?

This PR enables the exclusion of tests from TIAF via a JSON file. At the moment it is only possible to exclude tests using the tiaf.json file that is generated when building the project. This behaviour remains the default, only changing to read tests from a file if `--exclude=YOUR/PATH/HERE` is set in the C++ runtime arguments.

## How was this PR tested?

This PR was tested manually.

### CODE REVIEW

**Review by jonawals — COMMENTED** [review]
This is a great addition, I only have some minor suggestions and then it's GTG 👍

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntime.cpp [hunk 1]
```diff
@@ -244,6 +244,7 @@ namespace TestImpact
     Runtime::Runtime(
         RuntimeConfig&& config,
         AZStd::optional<RepoPath> dataFile,
+        AZStd::optional<AZStd::vector<AZStd::string>> testsToExclude,
```
[the remark this answers is not in the document]
Comment (jcurtisk) [reply Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntime.cpp]: Should the entire pipeline for this variable be converted to purely a vector then, and we initialise it as an empty vector if no valid data is found?

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp [hunk 2]
```diff
@@ -111,6 +114,16 @@ namespace TestImpact
             return ParsePathOption(OptionKeys[SequenceReportKey], cmd);
         }
 
+        AZStd::optional<AZStd::vector<AZStd::string>> ParseExcludeTestsFile(const AZ::CommandLine& cmd)
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp:117]: suggestion: Having an `optional` vector might be overkill here as we can simply check if the vector is `.empty()` to see if any strings were generated. That would basically be the 'empty optional' state and is likely more consistent with other APIs across the source code

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp [hunk 3]
```diff
@@ -333,6 +352,11 @@ namespace TestImpact
         return m_sequenceReportFile;
     }
 
+    const AZStd::optional<AZStd::vector<AZStd::string>>& CommandLineOptions::GetExcludeTests() const
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp:355]: suggestion: Same comment as before about removing the use of `AZStd::optional`

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptionsUtils.h [hunk 4]
```diff
@@ -115,4 +115,7 @@ namespace TestImpact
 
     //! Attempts to parse an option value in seconds.
     AZStd::optional<AZStd::chrono::milliseconds> ParseSecondsOption(const AZStd::string& optionName, const AZ::CommandLine& cmd);
+
+    //! Attempts to parse the file data into a JSON array of test names
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptionsUtils.h:119]: uber nit: Missing ending full-stop/period (`.`)

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntimeUtils.cpp [hunk 5]
```diff
@@ -14,6 +14,8 @@
 #include <Artifact/Factory/TestImpactBuildTargetDescriptorFactory.h>
 #include <Artifact/Static/TestImpactTargetDescriptorCompiler.h>
 
+#include <AZCore/JSON/document.h>
```
Comment (hultonha) [inline Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntimeUtils.cpp:17]: update: Looks like we can remove this include as nothing in this file has changed? Do we really need to add this?

**Review by hultonha — COMMENTED** [review]
Overall looks like a very solid change I just have a question about one of the API designs. I welcome your feedback and can approve after any discussion. Thanks!

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntimeUtils.cpp (hunk 5 above)
Comment (jcurtisk) [reply Code/Tools/TestImpactFramework/Runtime/Code/Source/TestImpactRuntimeUtils.cpp:17]: Will remove this.

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp (hunk 2 above)
Comment (jcurtisk) [reply Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp]: This has now been changed in the most recent push - I had already solved this but hadn't pushed my changes into the repo, apologies.

**Comment by jcurtisk** [comment]
> Overall looks like a very solid change I just have a question about one of the API designs. I welcome your feedback and can approve after any discussion. Thanks!

These changes have now been implemented, I forgot to push them somehow(hence why the commits look like they happened yesterday). Thanks for the review!

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp (hunk 2 above)
Comment (hultonha) [reply Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptions.cpp]: Ah nice one, thanks for following-up 👍

File: Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptionsUtils.h (hunk 4 above)
Comment (hultonha) [reply Code/Tools/TestImpactFramework/Frontend/Console/Code/Source/TestImpactCommandLineOptionsUtils.h]: Thanks 👍

**Review by hultonha — APPROVED** [review]
Nice update! 👍 The only thing left is do we want to add tests to verify `ParseExcludedTestTargetsFromFile`/`ParseExcludeTestsFile` do the right thing 🙂 (can be done as a separate PR but could be something to mull over)

**Comment by jcurtisk** [comment]
> Nice update! 👍 The only thing left is do we want to add tests to verify `ParseExcludedTestTargetsFromFile`/`ParseExcludeTestsFile` do the right thing 🙂 (can be done as a separate PR but could be something to mull over)

Definitely something to consider. I believe there are tests for the C++ runtime but they're currently disabled, so they could be added to that. I will discuss it with John on Monday and then add a ticket for it if needed.

**Pull request merged**